File: face_recognition_flutter/lib/models/class.dart
import 'user.dart';

class ClassData {
  final int id;
  final String name;
  final String? description;
  final int? teacherId;
  final String? teacherName;
  final bool isActive;
  final DateTime createdAt;
  final DateTime? updatedAt;
  final int studentCount;
  
  // Additional fields for detailed views
  final List<User>? students;
  final String? code; // Some classes might have codes

  ClassData({
    required this.id,
    required this.name,
    this.description,
    this.teacherId,
    this.teacherName,
    required this.isActive,
    required this.createdAt,
    this.updatedAt,
    required this.studentCount,
    this.students,
    this.code,
  });

  factory ClassData.fromJson(Map<String, dynamic> json) {
    return ClassData(
      id: int.tryParse(json['id'].toString()) ?? 0,
      name: json['name'] ?? '',
      description: json['description'] as String?,
      teacherId: json['teacher_id'] != null ? int.tryParse(json['teacher_id'].toString()) : null,
      teacherName: json['teacher_name'] as String?,
      isActive: json['is_active'] == 1 || json['is_active'] == true || json['status'] == true,
      createdAt: DateTime.tryParse(json['created_at'] ?? '') ?? DateTime.now(),
      updatedAt: json['updated_at'] != null 
          ? DateTime.tryParse(json['updated_at'] as String)
          : null,
      studentCount: int.tryParse(json['student_count'].toString()) ?? 0,
      students: json['students'] != null 
          ? (json['students'] as List).map((s) => User.fromJson(s)).toList()
          : null,
      code: json['code'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'description': description,
      'teacher_id': teacherId,
      'teacher_name': teacherName,
      'is_active': isActive,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt?.toIso8601String(),
      'student_count': studentCount,
      'students': students?.map((s) => s.toJson()).toList(),
      'code': code,
    };
  }

  ClassData copyWith({
    int? id,
    String? name,
    String? description,
    int? teacherId,
    String? teacherName,
    bool? isActive,
    DateTime? createdAt,
    DateTime? updatedAt,
    int? studentCount,
    List<User>? students,
    String? code,
  }) {
    return ClassData(
      id: id ?? this.id,
      name: name ?? this.name,
      description: description ?? this.description,
      teacherId: teacherId ?? this.teacherId,
      teacherName: teacherName ?? this.teacherName,
      isActive: isActive ?? this.isActive,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      studentCount: studentCount ?? this.studentCount,
      students: students ?? this.students,
      code: code ?? this.code,
    );
  }
}

/// Model for class-student relationship
class ClassStudent {
  final int id;
  final int classId;
  final int studentId;
  final String studentCode;
  final DateTime createdAt;
  
  // Related data
  final String? studentName;
  final String? className;
  final User? student;

  ClassStudent({
    required this.id,
    required this.classId,
    required this.studentId,
    required this.studentCode,
    required this.createdAt,
    this.studentName,
    this.className,
    this.student,
  });

  factory ClassStudent.fromJson(Map<String, dynamic> json) {
    return ClassStudent(
      id: json['id'] ?? 0,
      classId: json['class_id'] ?? 0,
      studentId: json['student_id'] ?? 0,
      studentCode: json['student_code'] ?? '',
      createdAt: DateTime.tryParse(json['created_at'] ?? '') ?? DateTime.now(),
      studentName: json['student_name'] as String?,
      className: json['class_name'] as String?,
      student: json['student'] != null ? User.fromJson(json['student']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'class_id': classId,
      'student_id': studentId,
      'student_code': studentCode,
      'created_at': createdAt.toIso8601String(),
      'student_name': studentName,
      'class_name': className,
      'student': student?.toJson(),
    };
  }
}

File: face_recognition_flutter/backendjs/src/models/Class.js
const db = require('../config/database');

class Class {
    constructor(data) {
        this.id = data.id;
        this.name = data.name;
        this.code = data.code;
        this.year = data.year;
        this.description = data.description;
        this.teacher_id = data.teacher_id;
        this.status = data.status || data.is_active;
        this.created_at = data.created_at;
        this.updated_at = data.updated_at;
    }

    // Tìm class theo ID
    static async findById(id) {
        try {
            const [rows] = await db.execute(
                'SELECT * FROM classes WHERE id = ? AND status = TRUE',
                [id]
            );
            return rows.length > 0 ? new Class(rows[0]) : null;
        } catch (error) {
            throw new Error(`Error finding class by ID: ${error.message}`);
        }
    }

    // Tạo class mới
    static async create(classData) {
        try {
            const { name, code, year, description, teacher_id, status = 'active' } = classData;
            
            const [result] = await db.execute(
                'INSERT INTO classes (name, code, year, description, teacher_id, status) VALUES (?, ?, ?, ?, ?, ?)',
                [name, code, year, description, teacher_id, status]
            );
            
            return await Class.findById(result.insertId);
        } catch (error) {
            throw new Error(`Error creating class: ${error.message}`);
        }
    }

    // Cập nhật class
    static async update(id, updateData) {
        try {
            const fields = [];
            const values = [];

            Object.keys(updateData).forEach(key => {
                if (updateData[key] !== undefined) {
                    fields.push(`${key} = ?`);
                    values.push(updateData[key]);
                }
            });

            if (fields.length === 0) {
                throw new Error('No fields to update');
            }

            values.push(id);
            const query = `UPDATE classes SET ${fields.join(', ')}, updated_at = NOW() WHERE id = ?`;
            
            await db.execute(query, values);
            return await Class.findById(id);
        } catch (error) {
            throw new Error(`Error updating class: ${error.message}`);
        }
    }

    // Xóa class (soft delete)
    static async delete(id) {
        try {
            await db.execute(
                'UPDATE classes SET status = FALSE, updated_at = NOW() WHERE id = ?',
                [id]
            );
            return true;
        } catch (error) {
            throw new Error(`Error deleting class: ${error.message}`);
        }
    }

    // Lấy tất cả classes với phân trang
    static async getAll(page = 1, limit = 10) {
        try {
            const offset = (page - 1) * limit;
            
            const [rows] = await db.execute(`
                SELECT c.*, u.full_name as teacher_name 
                FROM classes c 
                LEFT JOIN users u ON c.teacher_id = u.id 
                WHERE c.status  = TRUE 
                ORDER BY c.created_at DESC 
                LIMIT ? OFFSET ?
            `, [limit, offset]);

            const [countResult] = await db.execute(
                'SELECT COUNT(*) as total FROM classes WHERE status  = TRUE'
            );

            return {
                classes: rows.map(row => ({
                    ...new Class(row),
                    teacher_name: row.teacher_name
                })),
                total: countResult[0].total,
                page,
                limit,
                totalPages: Math.ceil(countResult[0].total / limit)
            };
        } catch (error) {
            throw new Error(`Error getting classes: ${error.message}`);
        }
    }

    // Lấy students trong class
    static async getStudents(classId) {
        try {
            const [rows] = await db.execute(`
                SELECT u.*, cs.student_code 
                FROM users u 
                JOIN class_students cs ON u.id = cs.student_id 
                WHERE cs.class_id = ? AND u.status = TRUE
                ORDER BY cs.student_code
            `, [classId]);

            return rows;
        } catch (error) {
            throw new Error(`Error getting class students: ${error.message}`);
        }
    }

    // Thêm student vào class
    static async addStudent(classId, studentId, studentCode) {
        try {
            await db.execute(
                'INSERT INTO class_students (class_id, student_id, student_code) VALUES (?, ?, ?)',
                [classId, studentId, studentCode]
            );
            return true;
        } catch (error) {
            throw new Error(`Error adding student to class: ${error.message}`);
        }
    }

    // Xóa student khỏi class
    static async removeStudent(classId, studentId) {
        try {
            await db.execute(
                'DELETE FROM class_students WHERE class_id = ? AND student_id = ?',
                [classId, studentId]
            );
            return true;
        } catch (error) {
            throw new Error(`Error removing student from class: ${error.message}`);
        }
    }
}

module.exports = Class;